File: MyApp.Application/Service/DiscountService.cs
using AutoMapper;
using MyApp.Application.Dto.Request;
using MyApp.Application.Dto.Response;
using MyApp.Application.Interface.Repository;
using MyApp.Application.Interface.Service;
using Myapp.Domain.Entity;
using Myapp.Domain.Exceptions;

namespace MyApp.Application.Service;

public class DiscountService(
    IDiscountRepository discountRepository,
    IMapper mapper) : IDiscountService
{
    public async Task<DiscountResponse> CreateDiscount(DiscountRequest request)
    {
        request.StartDate = DateTime.SpecifyKind(request.StartDate, DateTimeKind.Utc);
        request.EndDate = DateTime.SpecifyKind(request.EndDate, DateTimeKind.Utc);
        if (await discountRepository.ExistsByPercent(request.Percent, request.StartDate, request.EndDate))
            throw new AppException(ErrorCode.DISCOUNT_EXISTED);

        Discount discount= mapper.Map<Discount>(request);

        await discountRepository.Create(discount);
        return mapper.Map<DiscountResponse>(discount);
    }

    public async Task<DiscountResponse> UpdateDiscount(int id, DiscountRequest request)
    {
        request.StartDate = DateTime.SpecifyKind(request.StartDate, DateTimeKind.Utc);
        request.EndDate = DateTime.SpecifyKind(request.EndDate, DateTimeKind.Utc);
        var discount = await discountRepository.GetById(id)
                       ?? throw new AppException(ErrorCode.DISCOUNT_NOT_FOUND);

        mapper.Map(request, discount);

        await discountRepository.Update(discount);
        return mapper.Map<DiscountResponse>(discount);
    }

    public async Task DeleteDiscount(int id)
    {
        var discount = await discountRepository.GetById(id)
                       ?? throw new AppException(ErrorCode.DISCOUNT_NOT_FOUND);

        await discountRepository.Delete(discount);
    }

    public async Task<DiscountResponse> GetDiscountById(int id)
    {
        var discount = await discountRepository.GetById(id)
                       ?? throw new AppException(ErrorCode.DISCOUNT_NOT_FOUND);

        return mapper.Map<DiscountResponse>(discount);
    }

    public async Task<List<DiscountResponse>> GetAllDiscounts()
    {
        var discounts = await discountRepository.GetAll();
        return mapper.Map<List<DiscountResponse>>(discounts);
    }
}

File: fe-project-bigin/src/api/discountService.ts
import axiosClient from './axiosClient'
import type { ApiResponse } from './bookService'

export interface DiscountResponse {
  id: number;
  minApply: number;
  maxApply: number;
  percent: number;
  isActive: boolean;
  startDate: string; // ISO Date string
  endDate: string; // ISO Date string
}

export interface DiscountRequest {
  minApply: number;
  maxApply: number;
  percent: number;
  startDate: string;
  endDate: string;
}

export const discountService = {
  getAllDiscounts(): Promise<ApiResponse<DiscountResponse[]>> {
    return axiosClient.get('/api/discounts')
  },
  
  getDiscountById(id: number): Promise<ApiResponse<DiscountResponse>> {
    return axiosClient.get(`/api/discounts/${id}`)
  },

  createDiscount(data: DiscountRequest): Promise<ApiResponse<DiscountResponse>> {
    return axiosClient.post('/api/discounts', data)
  },

  updateDiscount(id: number, data: DiscountRequest): Promise<ApiResponse<DiscountResponse>> {
    return axiosClient.put(`/api/discounts/${id}`, data)
  },

  deleteDiscount(id: number): Promise<ApiResponse<string>> {
    return axiosClient.delete(`/api/discounts/${id}`)
  }
}
